### PR TITLE
UCT/IB: Introduce MD capability for non-blocking registration memory types

### PR DESCRIPTION
**Pull request by ivankochin**
## What
Add `uct_md_attr_v2_t::odp_reg_mem_types` to fill the ODPable transports for each MD.

## Why ?
ODPv2 support. We are planning to introduce UCP interface that forces ODP memory registration for certain MDs and therefore we want to check availability of ODP registration.

### CODE REVIEW

File: src/uct/api/v2/uct_v2.h [hunk 1]
```diff
@@ -671,7 +671,10 @@ typedef enum uct_md_attr_field {
     UCT_MD_ATTR_FIELD_EXPORTED_MKEY_PACKED_SIZE = UCS_BIT(13),
 
     /** Unique global identifier of the memory domain. */
-    UCT_MD_ATTR_FIELD_GLOBAL_ID                 = UCS_BIT(14)
+    UCT_MD_ATTR_FIELD_GLOBAL_ID                 = UCS_BIT(14),
+
+    /** Indicate memory types that are ODPable */
+    UCT_MD_ATTR_FIELD_ODP_REG_MEM_TYPES         = UCS_BIT(15)
```
Comment (brminich) [inline src/uct/api/v2/uct_v2.h:677]: let's rename it according to the API, e.g. `UCT_MD_ATTR_FIELD_NONBLOCK_REG_MEM_TYPES` or `UCT_MD_ATTR_FIELD_NB_REG_MEM_TYPES`
and the field itself as well

File: src/uct/api/v2/uct_v2.h [hunk 2]
```diff
@@ -671,7 +671,10 @@ typedef enum uct_md_attr_field {
     UCT_MD_ATTR_FIELD_EXPORTED_MKEY_PACKED_SIZE = UCS_BIT(13),
 
     /** Unique global identifier of the memory domain. */
-    UCT_MD_ATTR_FIELD_GLOBAL_ID                 = UCS_BIT(14)
+    UCT_MD_ATTR_FIELD_GLOBAL_ID                 = UCS_BIT(14),
+
+    /** Indicate memory types that are ODPable */
```
Comment (gleon99) [inline src/uct/api/v2/uct_v2.h:676]: Maybe rephrase to `Indicate memory types that the MD can register using ODP`, sounds more formal?
@brminich  WDYT?

Comment (ivankochin) [reply src/uct/api/v2/uct_v2.h:676]:
```suggestion
    /** Indicate memory types that the MD can register using ODP */
```

File: src/uct/api/v2/uct_v2.h [hunk 3]
```diff
@@ -772,6 +775,11 @@ typedef struct {
      * Memory Domains belong to the same device.
      */
     char              global_id[UCT_MD_GLOBAL_ID_MAX];
+
+    /**
+     * Bitmap of memory types that are ODPable.
```
Comment (ivankochin) [inline src/uct/api/v2/uct_v2.h:780]:
```suggestion
     * Bitmap of memory types that can be registered using ODP
```

File: src/uct/api/v2/uct_v2.h (hunk 2 above)
Comment (ivankochin) [reply src/uct/api/v2/uct_v2.h]: WDYTA?
```
/** Indicate memory types that are suitable for non-blocking registration */
```

Comment (gleon99) [reply src/uct/api/v2/uct_v2.h]: 👍🏼

File: src/uct/api/v2/uct_v2.h [hunk 4]
```diff
@@ -671,7 +671,10 @@ typedef enum uct_md_attr_field {
     UCT_MD_ATTR_FIELD_EXPORTED_MKEY_PACKED_SIZE = UCS_BIT(13),
 
     /** Unique global identifier of the memory domain. */
-    UCT_MD_ATTR_FIELD_GLOBAL_ID                 = UCS_BIT(14)
+    UCT_MD_ATTR_FIELD_GLOBAL_ID                 = UCS_BIT(14),
+
+    /** Indicate memory types that the MD can register using ODP */
```
Comment (ivankochin) [inline src/uct/api/v2/uct_v2.h:676]:
```suggestion
    /** Indicate memory types that are suitable for non-blocking registration */
```

File: src/uct/api/v2/uct_v2.h [hunk 5]
```diff
@@ -772,6 +775,11 @@ typedef struct {
      * Memory Domains belong to the same device.
      */
     char              global_id[UCT_MD_GLOBAL_ID_MAX];
+
+    /**
+     * Bitmap of memory types that can be registered using ODP
```
Comment (ivankochin) [inline src/uct/api/v2/uct_v2.h:780]:
```suggestion
     * Bitmap of memory types that are suitable for non-blocking registration
```

File: src/uct/api/v2/uct_v2.h [hunk 6]
```diff
@@ -772,6 +775,11 @@ typedef struct {
      * Memory Domains belong to the same device.
      */
     char              global_id[UCT_MD_GLOBAL_ID_MAX];
+
+    /**
+     * Bitmap of memory types that are suitable for non-blocking registration
+     */
+    uint64_t          nb_mem_types;
```
Comment (brminich) [inline src/uct/api/v2/uct_v2.h:782]: `nb_reg_mem_types`

Comment (ivankochin) [reply src/uct/api/v2/uct_v2.h:782]:
```suggestion
    uint64_t          nb_reg_mem_types;
```

File: src/uct/base/uct_md.c [hunk 7]
```diff
@@ -441,6 +441,8 @@ uct_md_attr_v2_copy(uct_md_attr_v2_t *dst, const uct_md_attr_v2_t *src)
                               UCT_MD_ATTR_FIELD_EXPORTED_MKEY_PACKED_SIZE);
     UCT_MD_ATTR_V2_FIELD_COPY(dst, src, global_id,
                               UCT_MD_ATTR_FIELD_GLOBAL_ID);
+    UCT_MD_ATTR_V2_FIELD_COPY(dst, src, nb_mem_types,
```
Comment (ivankochin) [inline src/uct/base/uct_md.c:444]:
```suggestion
    UCT_MD_ATTR_V2_FIELD_COPY(dst, src, nb_reg_mem_types,
```

File: src/uct/ib/base/ib_md.c [hunk 8]
```diff
@@ -282,6 +282,7 @@ static ucs_status_t uct_ib_md_query(uct_md_h uct_md, uct_md_attr_v2_t *md_attr)
     md_attr->access_mem_types          = UCS_BIT(UCS_MEMORY_TYPE_HOST);
     md_attr->detect_mem_types          = 0;
     md_attr->dmabuf_mem_types          = 0;
+    md_attr->nb_mem_types              = UCS_BIT(UCS_MEMORY_TYPE_HOST);
```
Comment (ivankochin) [inline src/uct/ib/base/ib_md.c:285]:
```suggestion
    md_attr->nb_reg_mem_types          = UCS_BIT(UCS_MEMORY_TYPE_HOST);
```

File: src/uct/api/v2/uct_v2.h (hunk 2 above)
Comment (brminich) [reply src/uct/api/v2/uct_v2.h]: 👍

**Review by brminich — APPROVED** [review]

**Comment by ivankochin** [comment]
@yosefe pls review

File: src/uct/api/v2/uct_v2.h [hunk 9]
```diff
@@ -671,7 +671,10 @@ typedef enum uct_md_attr_field {
     UCT_MD_ATTR_FIELD_EXPORTED_MKEY_PACKED_SIZE = UCS_BIT(13),
 
     /** Unique global identifier of the memory domain. */
-    UCT_MD_ATTR_FIELD_GLOBAL_ID                 = UCS_BIT(14)
+    UCT_MD_ATTR_FIELD_GLOBAL_ID                 = UCS_BIT(14),
+
+    /** Indicate memory types that are suitable for non-blocking registration */
+    UCT_MD_ATTR_FIELD_NB_REG_MEM_TYPES          = UCS_BIT(15)
```
Comment (yosefe) [inline src/uct/api/v2/uct_v2.h:677]: UCT_MD_ATTR_FIELD_REG_NONBLOCK_MEM_TYPES
reason: the flag is called UCT_MD_MEM_FLAG_NONBLOCK

File: src/uct/api/v2/uct_v2.h [hunk 10]
```diff
@@ -772,6 +775,11 @@ typedef struct {
      * Memory Domains belong to the same device.
      */
     char              global_id[UCT_MD_GLOBAL_ID_MAX];
+
+    /**
+     * Bitmap of memory types that are suitable for non-blocking registration
+     */
+    uint64_t          nb_reg_mem_types;
```
Comment (yosefe) [inline src/uct/api/v2/uct_v2.h:782]: reg_nonblock_mem_types

**Review by brminich — APPROVED** [review]

File: src/uct/api/v2/uct_v2.h [hunk 11]
```diff
@@ -671,7 +671,10 @@ typedef enum uct_md_attr_field {
     UCT_MD_ATTR_FIELD_EXPORTED_MKEY_PACKED_SIZE = UCS_BIT(13),
 
     /** Unique global identifier of the memory domain. */
-    UCT_MD_ATTR_FIELD_GLOBAL_ID                 = UCS_BIT(14)
+    UCT_MD_ATTR_FIELD_GLOBAL_ID                 = UCS_BIT(14),
+
+    /** Indicate memory types that are suitable for non-blocking registration */
```
Comment (yosefe) [inline src/uct/api/v2/uct_v2.h:676]: `Bitmap of memory types that support @ref UCT_MD_MEM_FLAG_NONBLOCK flag for @ref uct_md_mem_reg or @ref uct_md_mem_reg_v2`

File: src/uct/ib/base/ib_md.c [hunk 12]
```diff
@@ -282,6 +282,7 @@ static ucs_status_t uct_ib_md_query(uct_md_h uct_md, uct_md_attr_v2_t *md_attr)
     md_attr->access_mem_types          = UCS_BIT(UCS_MEMORY_TYPE_HOST);
     md_attr->detect_mem_types          = 0;
     md_attr->dmabuf_mem_types          = 0;
+    md_attr->reg_nonblock_mem_types    = UCS_BIT(UCS_MEMORY_TYPE_HOST);
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: 1. it's not really supported on devx
2. what about other md's? any md that has dummy registration or registration without pinning should support it.

File: src/uct/api/v2/uct_v2.h [hunk 13]
```diff
@@ -772,6 +778,11 @@ typedef struct {
      * Memory Domains belong to the same device.
      */
     char              global_id[UCT_MD_GLOBAL_ID_MAX];
+
+    /**
+     * Bitmap of memory types that are suitable for non-blocking registration
+     */
+    uint64_t          reg_nonblock_mem_types;
```
Comment (yosefe) [inline src/uct/api/v2/uct_v2.h:785]: maybe add this field, and its corresponding bit, after the field "reg_mem_types" ?

File: src/uct/sm/mm/base/mm_md.c [hunk 14]
```diff
@@ -69,14 +69,15 @@ ucs_status_t uct_mm_seg_new(void *address, size_t length, uct_mm_seg_t **seg_p)
...
-    md_attr->dmabuf_mem_types = 0;
+    md_attr->flags                  = UCT_MD_FLAG_RKEY_PTR | UCT_MD_FLAG_NEED_RKEY;
+    md_attr->max_reg                = 0;
+    md_attr->max_alloc              = 0;
+    md_attr->alloc_mem_types        = UCS_BIT(UCS_MEMORY_TYPE_HOST);
+    md_attr->cache_mem_types        = UCS_BIT(UCS_MEMORY_TYPE_HOST);
+    md_attr->access_mem_types       = UCS_BIT(UCS_MEMORY_TYPE_HOST);
+    md_attr->detect_mem_types       = 0;
+    md_attr->dmabuf_mem_types       = 0;
+    md_attr->reg_nonblock_mem_types = 0;
```
Comment (yosefe) [inline src/uct/sm/mm/base/mm_md.c:80]: why need to initialize this, and not md_attr->reg_mem_types?

File: src/uct/api/v2/uct_v2.h (hunk 13 above)
Comment (brminich) [reply src/uct/api/v2/uct_v2.h:785]: this would break combatibility

File: src/uct/sm/mm/base/mm_md.c (hunk 14 above)
Comment (brminich) [reply src/uct/sm/mm/base/mm_md.c:80]: i think we do not need both, because `md_attr->flags` does not include `UCT_MD_FLAG_REG`

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c [hunk 15]
```diff
@@ -769,6 +769,7 @@ uct_ib_mlx5_devx_check_odp(uct_ib_mlx5_md_t *md,
         } else {
             md->super.config.odp.max_size = 1ul << 28;
         }
+        md->super.reg_nonblock_mem_types = UCS_BIT(UCS_MEMORY_TYPE_HOST);
```
Comment (brminich) [inline src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:772]: please initialize it explicitly to `0` in some place if no odp support

File: src/uct/api/v2/uct_v2.h (hunk 13 above)
Comment (yosefe) [reply src/uct/api/v2/uct_v2.h:785]: v2 does not need to be compatible

File: src/uct/sm/mm/base/mm_md.c (hunk 14 above)
Comment (yosefe) [reply src/uct/sm/mm/base/mm_md.c:80]: right

**Comment by yosefe** [comment]
@shamisp @tonycurtis can you pls review the UCT API change?

File: src/uct/ib/base/ib_md.c [hunk 16]
```diff
@@ -282,6 +282,7 @@ static ucs_status_t uct_ib_md_query(uct_md_h uct_md, uct_md_attr_v2_t *md_attr)
     md_attr->access_mem_types          = UCS_BIT(UCS_MEMORY_TYPE_HOST);
     md_attr->detect_mem_types          = 0;
     md_attr->dmabuf_mem_types          = 0;
+    md_attr->reg_nonblock_mem_types    = md->reg_nonblock_mem_types;
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c:285]: swap with next line

File: src/uct/ib/base/ib_md.c [hunk 17]
```diff
@@ -2058,6 +2059,7 @@ static ucs_status_t uct_ib_verbs_md_open(struct ibv_device *ibv_device,
         goto err_md_free;
     }
 
+    md->reg_nonblock_mem_types = 0;
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c:2062]: maybe move it to uct_ib_md_open_common, along with the initialization of  md->reg_mem_types ?

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c [hunk 18]
```diff
@@ -769,6 +769,9 @@ uct_ib_mlx5_devx_check_odp(uct_ib_mlx5_md_t *md,
...
```
Comment (yosefe) [inline src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:772]: space line before

File: src/uct/ib/base/ib_md.c (hunk 17 above)
Comment (ivankochin) [reply src/uct/ib/base/ib_md.c:2062]: If I understand the logic correctly, then moving it to `uct_ib_md_open_common` will overwrite value that can be set during `uct_ib_mlx5_devx_check_odp`.

File: src/uct/api/v2/uct_v2.h [hunk 19]
```diff
@@ -640,38 +640,44 @@ typedef enum uct_md_attr_field {
     /** Indicate memory types that the MD can register. */
     UCT_MD_ATTR_FIELD_REG_MEM_TYPES             = UCS_BIT(3),
 
+    /**
+     * Bitmap of memory types that support @ref UCT_MD_MEM_FLAG_NONBLOCK
```
[the remark this answers is not in the document]
Comment (ivankochin) [reply src/uct/api/v2/uct_v2.h:644]:
```suggestion
     * Indicate memory types that are suitable for non-blocking registration
```

File: src/uct/api/v2/uct_v2.h [hunk 20]
```diff
@@ -640,38 +640,44 @@ typedef enum uct_md_attr_field {
     /** Indicate memory types that the MD can register. */
     UCT_MD_ATTR_FIELD_REG_MEM_TYPES             = UCS_BIT(3),
 
+    /**
+     * Bitmap of memory types that support @ref UCT_MD_MEM_FLAG_NONBLOCK 
+     * flag for @ref uct_md_mem_reg or @ref uct_md_mem_reg_v2
+     */
```
Comment (ivankochin) [inline src/uct/api/v2/uct_v2.h:646]:
```suggestion
    /** Indicate memory types that are suitable for non-blocking registration. */
```

File: src/uct/ib/base/ib_md.c (hunk 17 above)
Comment (yosefe) [reply src/uct/ib/base/ib_md.c:2062]: ok
but probably we don't need it anyway since md is allocated by calloc

Comment (brminich) [reply src/uct/ib/base/ib_md.c:2062]: explicit initialization is a good "code documentation". BTW, do we really need to call `uct_ib_md_open_common` so late? Would it be possible to call it earlier, so that it can initialize common stuff?

Comment (yosefe) [reply src/uct/ib/base/ib_md.c:2062]: IMO we can call it earlier from mlx5 md init code (devx and mlx5dv) but I would not do it in this PR since it may cause unexpected issues, and needs more testing.
we already rely on md zero initialization for some other fields, AFAIR we discussed it with @Artemy-Mellanox

Comment (yosefe) [reply src/uct/ib/base/ib_md.c:2062]: IMO we should keep relying on calloc for now and maybe refactor uct_ib_md_open_common to be called earlier. for example seems it does not rely on the initialization of ODP flags.

Comment (brminich) [reply src/uct/ib/base/ib_md.c:2062]: ok, let's remove initialization in this PR and (maybe) refactor md open later in another PR

**Review by yosefe — APPROVED** [review]

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@shamisp @tonycurtis can you pls review the UCT API change?

File: src/uct/api/v2/uct_v2.h [hunk 21]
```diff
@@ -659,38 +659,41 @@ typedef enum uct_md_attr_field {
     /** Indicate memory types that the MD can register. */
     UCT_MD_ATTR_FIELD_REG_MEM_TYPES             = UCS_BIT(3),
 
+    /** Indicate memory types that are suitable for non-blocking registration. */
+    UCT_MD_ATTR_FIELD_REG_NONBLOCK_MEM_TYPES    = UCS_BIT(4),
+
```
[the remark this answers is not in the document]
Comment (brminich) [reply src/uct/api/v2/uct_v2.h:664]: @yosefe

[the remark this answers is not in the document]
Comment (shamisp) [reply src/uct/api/v2/uct_v2.h:664]: ok.

**Comment by shamisp** [comment]
👍
